File: common/src/main/scala/core/concurrentdemo/CountDownLatchDemo.scala
package core.concurrentdemo

import java.util.concurrent.{CountDownLatch, Executors}

/**
  * Created by yidxue on 2018/3/5
  */
object CountDownLatchDemo {

  def main(args: Array[String]): Unit = {
    val executor = Executors.newCachedThreadPool
    val latch = new CountDownLatch(3)
    val w1 = new Worker(latch, "张三")
    val w2 = new Worker(latch, "李四")
    val w3 = new Worker(latch, "王二")

    val boss = new Boss(latch)

    executor.execute(w3)
    executor.execute(w2)
    executor.execute(w1)
    executor.execute(boss)

    executor.shutdown()
  }
}

File: common/src/main/java/workflow/concurrentdemo/CountDownLatchDemo.java
package workflow.concurrentdemo;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * @author yidxue
 */
public class CountDownLatchDemo {

    public static void main(String[] args) {
        ExecutorService executor = Executors.newCachedThreadPool();

        CountDownLatch latch = new CountDownLatch(3);

        Worker w1 = new Worker(latch,"张三");
        Worker w2 = new Worker(latch,"李四");
        Worker w3 = new Worker(latch,"王二");

        Boss boss = new Boss(latch);

        // NOTICE submit 和 execute有什么区别
        executor.execute(w3);
        executor.execute(w2);
        executor.execute(w1);
        executor.execute(boss);

        executor.shutdown();
    }
}